stream-chat-android-client/src/main/java/io/getstream/chat/android/client/events/caching/TypingEventCache.kt

```diff
@@ -0,0 +1,234 @@
+    private val onEventFired: (event: ChatEvent) -> Unit,
+) : ChatEventCache {
+
+    /**
+     *  A list of currently typing users. Only [SelfStoppingTypingEvent]s
+     *  are stored here, which are by design self "cleaning" [TypingStartEvent]s.
+     *
+     *  [TypingStopEvent] are not stored, but their processing removes their
+     *  [SelfStoppingTypingEvent] counterpart.
+     */
+    private val _currentlyTypingUsers = mutableMapOf<TypingEventKeyValue, SelfStoppingTypingEvent>()
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    val currentlyTypingUsers: Map<TypingEventKeyValue, SelfStoppingTypingEvent> = _currentlyTypingUsers
+
+    /**
+     * Processes the incoming chat event accordingly.
+     *
+     * @param event The chat event to be processed.
+     */
+    override fun processEvent(event: ChatEvent) {
+        when (event) {
+            is TypingStartEvent -> {
+                processTypingStartEvent(event)
+            }
+            is TypingStopEvent -> {
+                processTypingStopEvent(event)
+            }
+            else -> {
+                throw IllegalArgumentException(
+                    "This class should be only used to cache typing events. " +
+                        "All other events will remain unprocessed."
+                )
+            }
+        }
+    }
+
+    /**
+     * Processes chat events by stopping the old "cleaning" job held
+     * by the previous value stored in the map under the same key
+     * and starts a new timed "cleaning" job.
+     *
+     * @param typingStartEvent The typing start event to be processed.
+     */
+    private fun processTypingStartEvent(typingStartEvent: TypingStartEvent) {
+        val keyValue = typingStartEvent.toStartTypingEventKeyValue()
+
+        // Cancel the previous timed death
+        _currentlyTypingUsers.getOrDefault(keyValue, null)?.cancelJob()
+
+        _currentlyTypingUsers[keyValue] = SelfStoppingTypingEvent(
+            coroutineScope = coroutineScope,
+            typingStartEvent = typingStartEvent,
+            delayTimeMs = cleanStaleEventMs
+        ) { typingStopEvent -> processTypingStopEvent(typingStopEvent) }
+
+        // Signal that the event has been processed
+        onEventFired(typingStartEvent)
+    }
+
+    /**
+     * Cancels the old "cleaning" job held by the previous
+     * value stored in the map under the same key and removes it.
+     *
+     * @param typingStopEvent The typing stop event to be processed.
+     */
+    private fun processTypingStopEvent(typingStopEvent: TypingStopEvent) {
+        val keyValue = typingStopEvent.toStartTypingEventKeyValue()
+
+        // Cancel the previous timed death
+        _currentlyTypingUsers.getOrDefault(keyValue, null)?.cancelJob()
+
+        _currentlyTypingUsers.remove(keyValue)
+
+        // Signal that the event has been processed
+        onEventFired(typingStopEvent)
+    }
+
+    /**
+     * Cancels [coroutineScope].
+     */
+    fun cancel() {
```

Comment:
It might not be needed if we decide to pass the same scope we are using in `OfflinePlugin` as it's already being canceled on user disconnect but let's wait for @kanat  input